File: api/app/routers/personal_services.py
"""Browse/detail endpoints for the personal radio services that share the
generic ULS record layout: GMRS, Aircraft (Part 87) and Ship (Part 80).

These three services are deliberately served by ONE config-driven router
factory rather than three near-identical hand-written modules. The goal for
this feature was explicit -- users of these services should not be
second-class citizens next to Amateur -- and the surest way to guarantee
that is to make feature parity structural: every service automatically gets
the same rate limiting, the same allow-listed sorting, the same partial-match
filtering, and the same detail-page sections, because there is only one
implementation of each. Adding a fourth service is then a dict entry, not a
new file to keep in sync.

(Amateur keeps its own module: it has service-specific concepts -- operator
class, trustee/club relationships, the New Hams celebration -- that don't
generalize, and rewriting a live, working endpoint to fit a shared abstraction
would risk real user-facing data for no functional gain.)
"""
from dataclasses import dataclass, field

from fastapi import APIRouter, Depends, HTTPException, Query, Request
from psycopg import AsyncConnection

from ..config import settings
from ..db import get_db
from ..history_codes import describe_history_code
from ..pagination import Page, PageParams, resolve_sort
from ..ratelimit import enforce_rate_limit


@dataclass(frozen=True)
class DetailSection:
    """An extra service-specific block returned by the detail endpoint."""

    key: str
    table: str
    # True when a subject legitimately has several rows (ship voyage
    # descriptions, which FCC splits across sequence-numbered rows).
    many: bool = False
    order_by: str | None = None


@dataclass(frozen=True)
class ServiceConfig:
    name: str
    prefix: str
    subject_type: str
    hd_table: str
    en_table: str
    hs_table: str
    # Extra JOIN applied to the browse query, aliased so browse_columns and
    # sortable_columns can reference it.
    browse_join: str = ""
    browse_columns: str = ""
    sortable_columns: dict[str, str] = field(default_factory=dict)
    # Extra browse filters: query-param name -> (SQL condition, kind).
    # kind "ilike" does partial matching; "exact" compares uppercased.
    extra_filters: dict[str, tuple[str, str]] = field(default_factory=dict)
    detail_sections: tuple[DetailSection, ...] = ()


BASE_SORTABLE = {
    "call_sign": "hd.call_sign",
    "license_status": "hd.license_status",
    "entity_name": "en.entity_name",
    "state": "en.state",
    "city": "en.city",
    "grant_date": "hd.grant_date",
    "expired_date": "hd.expired_date",
}

SERVICE_CONFIGS: dict[str, ServiceConfig] = {
    "gmrs": ServiceConfig(
        name="gmrs",
        prefix="/api/gmrs",
        subject_type="gmrs_license",
        hd_table="gmrs_hd",
        en_table="gmrs_en",
        hs_table="gmrs_hs",
        sortable_columns=dict(BASE_SORTABLE),
    ),
    "aircraft": ServiceConfig(
        name="aircraft",
        prefix="/api/aircraft",
        subject_type="aircraft_license",
        hd_table="aircr_hd",
        en_table="aircr_en",
        hs_table="aircr_hs",
        browse_join="LEFT JOIN aircr_ac ac ON ac.unique_system_identifier = hd.unique_system_identifier",
        browse_columns="ac.n_number, ac.type_of_carrier, ac.fleet_indicator",
        sortable_columns={
            **BASE_SORTABLE,
            "n_number": "ac.n_number",
            "type_of_carrier": "ac.type_of_carrier",
        },
        # The FAA tail number is how aircraft operators actually search.
        extra_filters={"n_number": ("ac.n_number ILIKE %(n_number)s", "ilike")},
        detail_sections=(DetailSection(key="aircraft_specific", table="aircr_ac"),),
    ),
    "ship": ServiceConfig(
        name="ship",
        prefix="/api/ship",
        subject_type="ship_license",
        hd_table="ship_hd",
        en_table="ship_en",
        hs_table="ship_hs",
        browse_join="LEFT JOIN ship_sh sh ON sh.unique_system_identifier = hd.unique_system_identifier",
        browse_columns="sh.ship_name, sh.general_class, sh.special_class, sh.station_number",
        sortable_columns={
            **BASE_SORTABLE,
            "ship_name": "sh.ship_name",
            "general_class": "sh.general_class",
        },
        extra_filters={
            "ship_name": ("sh.ship_name ILIKE %(ship_name)s", "ilike"),
            # station_number is the vessel's MMSI -- a primary real-world
            # identifier, so it gets its own filter.
            "mmsi": ("sh.station_number ILIKE %(mmsi)s", "ilike"),
        },
        detail_sections=(
            DetailSection(key="ship_specific", table="ship_sh"),
            DetailSection(key="radio_equipment", table="ship_sr"),
            DetailSection(
                key="voyages", table="ship_sv", many=True,
                # FCC splits one long description across sequence-numbered
                # rows; they must come back in order to read correctly.
                order_by="voyage_number",
            ),
            DetailSection(key="exemption_request", table="ship_se"),
        ),
    ),
}


def build_router(cfg: ServiceConfig) -> APIRouter:
    router = APIRouter(prefix=cfg.prefix, tags=[cfg.name])
    sortable = cfg.sortable_columns
    extra_cols = f", {cfg.browse_columns}" if cfg.browse_columns else ""

    @router.get("", response_model=Page)
    async def browse(
        request: Request,
        callsign: str | None = Query(None),
        name: str | None = Query(None),
        city: str | None = Query(None),
        state: str | None = Query(None),
        status_code: str | None = Query(None, alias="status"),
        n_number: str | None = Query(None),
        ship_name: str | None = Query(None),
        mmsi: str | None = Query(None),
        sort: str | None = Query(None, description=f"One of {sorted(sortable)}"),
        order: str | None = Query(None, description="asc or desc"),
        page_params: PageParams = Depends(),
        conn: AsyncConnection = Depends(get_db),
    ):
        sort_expr, sort_direction = resolve_sort(
            sort, order, sortable, default_column="call_sign"
        )
        client_ip = request.client.host if request.client else "unknown"
        await enforce_rate_limit(
            f"{cfg.name}-browse:{client_ip}",
            settings.rate_limit_search_max,
            settings.rate_limit_search_window_seconds,
        )

        conditions: list[str] = []
        params: dict = {}
        if callsign:
            conditions.append("hd.call_sign ILIKE %(callsign)s")
            params["callsign"] = f"%{callsign.strip()}%"
        if name:
            conditions.append(
                "(en.entity_name ILIKE %(name)s OR en.first_name ILIKE %(name)s"
                " OR en.last_name ILIKE %(name)s)"
            )
            params["name"] = f"%{name.strip()}%"
        if city:
            conditions.append("en.city ILIKE %(city)s")
            params["city"] = f"%{city.strip()}%"
        if state:
            conditions.append("en.state ILIKE %(state)s")
            params["state"] = f"%{state.strip()}%"
        if status_code:
            conditions.append("hd.license_status = %(status_code)s")
            params["status_code"] = status_code.upper()

        # Service-specific filters, ignored when the caller passes one that
        # doesn't apply to this service (e.g. ?ship_name= against /api/gmrs).
        supplied = {"n_number": n_number, "ship_name": ship_name, "mmsi": mmsi}
        for key, (condition, kind) in cfg.extra_filters.items():
            value = supplied.get(key)
            if not value:
                continue
            conditions.append(condition)
            params[key] = f"%{value.strip()}%" if kind == "ilike" else value.strip().upper()

        where_clause = f"WHERE {' AND '.join(conditions)}" if conditions else ""
        from_clause = f"""
            FROM {cfg.hd_table} hd
            LEFT JOIN {cfg.en_table} en ON en.unique_system_identifier = hd.unique_system_identifier
            {cfg.browse_join}
        """

        async with conn.cursor() as cur:
            await cur.execute(
                f"SELECT count(*) AS total {from_clause} {where_clause}", params
            )
            total = (await cur.fetchone())["total"]

            await cur.execute(
                f"""
                SELECT hd.call_sign, hd.license_status, hd.radio_service_code,
                       hd.grant_date, hd.expired_date,
                       en.entity_name, en.state, en.city{extra_cols}
                {from_clause}
                {where_clause}
                ORDER BY {sort_expr} {sort_direction} NULLS LAST, hd.call_sign
                LIMIT %(limit)s OFFSET %(offset)s
                """,
                {**params, "limit": page_params.limit, "offset": page_params.offset},
            )
            rows = await cur.fetchall()

        return Page(
            items=rows, page=page_params.page,
            page_size=page_params.page_size, total=total,
        )

    @router.get("/{call_sign}")
    async def detail(
        request: Request,
        call_sign: str,
        conn: AsyncConnection = Depends(get_db),
    ):
        call_sign = call_sign.upper()
        client_ip = request.client.host if request.client else "unknown"
        await enforce_rate_limit(
            f"{cfg.name}-detail:{client_ip}",
            settings.rate_limit_search_max,
            settings.rate_limit_search_window_seconds,
        )

        async with conn.cursor() as cur:
            # A callsign can appear under several unique_system_identifiers
            # over time (reassigned after a prior licence lapsed). Resolve
            # the CURRENT holder -- prefer an active record, else the most
            # recently granted -- and scope every other section to that same
            # identifier so two holders' data is never blended.
            await cur.execute(
                f"""
                SELECT unique_system_identifier FROM {cfg.hd_table}
                WHERE call_sign = %s
                ORDER BY (license_status = 'A') DESC, grant_date DESC NULLS LAST
                LIMIT 1
                """,
                (call_sign,),
            )
            row = await cur.fetchone()
            if row is None:
                raise HTTPException(status_code=404, detail="Callsign not found")
            usid = row["unique_system_identifier"]

            await cur.execute(
                f"SELECT * FROM {cfg.hd_table} WHERE unique_system_identifier = %s", (usid,)
            )
            header = await cur.fetchone()

            await cur.execute(
                f"SELECT * FROM {cfg.en_table} WHERE unique_system_identifier = %s", (usid,)
            )
            entity = await cur.fetchone()

            sections: dict = {}
            for section in cfg.detail_sections:
                order_clause = f" ORDER BY {section.order_by}" if section.order_by else ""
                await cur.execute(
                    f"SELECT * FROM {section.table}"
                    f" WHERE unique_system_identifier = %s{order_clause}",
                    (usid,),
                )
                sections[section.key] = (
                    await cur.fetchall() if section.many else await cur.fetchone()
                )

            # Full history across ALL holders of the callsign, so the
            # timeline shows a prior holder's expiry then the new grant.
            await cur.execute(
                f"SELECT * FROM {cfg.hs_table} WHERE callsign = %s"
                " ORDER BY log_date DESC NULLS LAST",
                (call_sign,),
            )
            history = await cur.fetchall()
            for entry in history:
                entry["code_description"] = describe_history_code(entry.get("code"))

            await cur.execute(
                """
                SELECT field_name, old_value, new_value, source_file,
                       effective_date, detected_at
                FROM change_events
                WHERE subject_type = %s AND subject_key = %s
                ORDER BY detected_at DESC
                LIMIT 100
                """,
                (cfg.subject_type, call_sign),
            )
            change_log = await cur.fetchall()

            # Cross-service identity grouping: the headline reason these
            # services are worth adding together, since one FRN commonly
            # holds e.g. both an Amateur and a GMRS licence.
            related = []
            if entity and entity.get("frn"):
                await cur.execute(
                    """
                    SELECT source, subject_key, entity_name, licensee_id
                    FROM identity_by_frn
                    WHERE frn = %s AND subject_key <> %s
                    """,
                    (entity["frn"], call_sign),
                )
                related = await cur.fetchall()

        return {
            "service": cfg.name,
            "header": header,
            "entity": entity,
            "history": history,
            "change_log": change_log,
            "related_identities": related,
            **sections,
        }

    return router


routers = [build_router(cfg) for cfg in SERVICE_CONFIGS.values()]

File: web/src/lib/personalServices.js
/**
 * Display configuration for the personal radio services that share the
 * generic ULS record layout: GMRS, Aircraft (Part 87) and Ship (Part 80).
 *
 * This mirrors the API's own config-driven router factory
 * (api/app/routers/personal_services.py). The reason is the same on both
 * sides: the explicit goal of adding these services was that their users
 * should not be second-class citizens next to Amateur, and the surest way
 * to guarantee that is to make parity structural. There is ONE browse
 * component and ONE detail component; every service automatically gets the
 * same sorting, filtering, tooltips, code definitions, watch links and
 * cross-service identity grouping, because there is only one implementation
 * of each. Adding a fourth service is an entry in this file, not a new set
 * of pages to keep in sync.
 *
 * Every coded/abbreviated column below names a `category` from
 * fieldDefs.js, so it renders with a definition (inline parenthetical when
 * short, tooltip when long) exactly as Amateur and Tower fields do. That is
 * a project-wide standard, not a per-page choice.
 */

// Status options shared by every ULS service (the HD record's
// license_status is a generic ULS field, not an Amateur-specific one).
const STATUS_OPTIONS = [
	{ value: '', label: 'Any status' },
	{ value: 'A', label: 'Active' },
	{ value: 'E', label: 'Expired' },
	{ value: 'C', label: 'Cancelled' },
	{ value: 'T', label: 'Terminated' }
];

// Filters every service supports, because they come from the shared
// HD/EN records.
const COMMON_FILTERS = [
	{ key: 'callsign', placeholder: 'Callsign (partial)' },
	{ key: 'name', placeholder: 'Licensee name (partial)' },
	{ key: 'city', placeholder: 'City (partial)' },
	{ key: 'state', placeholder: 'State', maxlength: 2 },
	{ key: 'status', kind: 'select', options: STATUS_OPTIONS }
];

// Columns every service shows. Service-specific columns are spliced in
// after the callsign/status pair.
const COMMON_TAIL_COLUMNS = [
	{ key: 'entity_name', label: 'Licensee', sort: 'entity_name' },
	{ key: 'location', label: 'Location', sort: 'city', kind: 'location' },
	{ key: 'grant_date', label: 'Grant', sort: 'grant_date' },
	{ key: 'expired_date', label: 'Expires', sort: 'expired_date' }
];

const COMMON_HEAD_COLUMNS = [
	{ key: 'call_sign', label: 'Callsign', sort: 'call_sign', kind: 'callsign' },
	{ key: 'license_status', label: 'Status', sort: 'license_status', kind: 'status' }
];

// The License and Licensee blocks are identical across services because
// the underlying HD and EN records are byte-identical FCC record types.
const LICENSE_SECTION = {
	source: 'header',
	title: 'License',
	fields: [
		{ key: 'license_status', label: 'Status', category: 'license_status' },
		{ key: 'radio_service_code', label: 'Radio Service Code', category: 'radio_service_code' },
		{ key: 'uls_file_number', label: 'ULS File Number' },
		{ key: 'unique_system_identifier', label: 'ULS System ID' },
		{ key: 'grant_date', label: 'Grant Date' },
		{ key: 'expired_date', label: 'Expires' },
		{ key: 'effective_date', label: 'Effective Date' },
		{ key: 'last_action_date', label: 'Last Action Date' },
		{ key: 'cancellation_date', label: 'Cancellation Date' }
	]
};

const LICENSEE_SECTION = {
	source: 'entity',
	title: 'Licensee',
	kind: 'entity'
};

export const PERSONAL_SERVICES = {
	gmrs: {
		name: 'gmrs',
		label: 'GMRS',
		title: 'GMRS Licenses',
		heading: 'General Mobile Radio Service (GMRS)',
		blurb:
			'GMRS is a licensed two-way radio service for short-distance personal and family ' +
			'communication. One licence covers the licensee and their immediate family.',
		apiPath: '/gmrs',
		route: '/gmrs',
		filters: COMMON_FILTERS,
		columns: [...COMMON_HEAD_COLUMNS, ...COMMON_TAIL_COLUMNS],
		// GMRS has no service-specific FCC record -- HD/EN/HS is the whole
		// dataset -- so its detail page is the shared sections only.
		detailSections: [LICENSE_SECTION, LICENSEE_SECTION]
	},

	aircraft: {
		name: 'aircraft',
		label: 'Aircraft',
		title: 'Aircraft Licenses',
		heading: 'Aircraft Radio Stations (Part 87)',
		blurb:
			'Radio station licences for aircraft. Most flying in the US does not require one, ' +
			'but international flights and some commercial operations do.',
		apiPath: '/aircraft',
		route: '/aircraft',
		filters: [...COMMON_FILTERS, { key: 'n_number', placeholder: 'N-number / tail number' }],
		columns: [
			...COMMON_HEAD_COLUMNS,
			{ key: 'n_number', label: 'N-Number', sort: 'n_number', help: 'n_number' },
			{
				key: 'type_of_carrier',
				label: 'Carrier',
				sort: 'type_of_carrier',
				kind: 'code',
				category: 'type_of_carrier'
			},
			...COMMON_TAIL_COLUMNS
		],
		detailSections: [
			LICENSE_SECTION,
			LICENSEE_SECTION,
			{
				source: 'aircraft_specific',
				title: 'Aircraft Details',
				fields: [
					{ key: 'n_number', label: 'N-Number (tail number)' },
					{ key: 'aircraft_count', label: 'Aircraft Count' },
					{ key: 'type_of_carrier', label: 'Type of Carrier', category: 'type_of_carrier' },
					{ key: 'portable_indicator', label: 'Portable', category: 'yes_no' },
					{ key: 'fleet_indicator', label: 'Fleet', category: 'yes_no' }
				]
			}
		]
	},

	ship: {
		name: 'ship',
		label: 'Ship',
		title: 'Ship Licenses',
		heading: 'Ship Radio Stations (Part 80)',
		blurb:
			'Radio station licences for vessels. Recreational boats in US waters usually do not ' +
			'need one, but international voyages and compulsorily-equipped ships do.',
		apiPath: '/ship',
		route: '/ship',
		filters: [
			...COMMON_FILTERS,
			{ key: 'ship_name', placeholder: 'Ship name (partial)' },
			{ key: 'mmsi', placeholder: 'MMSI number' }
		],
		columns: [
			...COMMON_HEAD_COLUMNS,
			{ key: 'ship_name', label: 'Ship Name', sort: 'ship_name' },
			{
				key: 'general_class',
				label: 'Class',
				sort: 'general_class',
				kind: 'code',
				category: 'ship_general_class'
			},
			{ key: 'station_number', label: 'MMSI', help: 'station_number' },
			...COMMON_TAIL_COLUMNS
		],
		detailSections: [
			LICENSE_SECTION,
			LICENSEE_SECTION,
			{
				source: 'ship_specific',
				title: 'Ship Details',
				fields: [
					{ key: 'ship_name', label: 'Ship Name' },
					{ key: 'station_number', label: 'MMSI Number' },
					{
						key: 'type_of_authorization',
						label: 'Type of Authorization',
						category: 'type_of_authorization'
					},
					{ key: 'general_class', label: 'General Class', category: 'ship_general_class' },
					{ key: 'special_class', label: 'Special Class', category: 'ship_special_class' },
					{ key: 'ship_number', label: 'Official Number of Ship' },
					{ key: 'gross_tonnage', label: 'Gross Tonnage' },
					{ key: 'ship_length', label: 'Ship Length' },
					{ key: 'count_in_fleet', label: 'Count in Fleet' },
					{ key: 'international_voyages', label: 'International Voyages', category: 'yes_no' },
					{ key: 'foreign_communications', label: 'Foreign Communications', category: 'yes_no' },
					{ key: 'radiotelegraph', label: 'Radiotelegraph Working Series' },
					{ key: 'mmsi_request', label: 'MMSI Requested', category: 'yes_no' },
					{ key: 'self_id_number', label: 'Sel Call Number' },
					{ key: 'comsat_id_number', label: 'Sel Call — INMARSAT' },
					{ key: 'working_freq_s1', label: 'Working Frequency 1' },
					{ key: 'working_freq_s2', label: 'Working Frequency 2' }
				]
			},
			{
				source: 'radio_equipment',
				title: 'Radio Equipment',
				fields: [
					{ key: 'vhf', label: 'VHF', category: 'yes_no' },
					{ key: 'mf', label: 'MF', category: 'yes_no' },
					{ key: 'hf', label: 'HF', category: 'yes_no' },
					{ key: 'dsc', label: 'DSC', category: 'yes_no' },
					{ key: 'inmarsat_a', label: 'INMARSAT A', category: 'yes_no' },
					{ key: 'inmarsat_b', label: 'INMARSAT B', category: 'yes_no' },
					{ key: 'inmarsat_c', label: 'INMARSAT C', category: 'yes_no' },
					{ key: 'inmarsat_m', label: 'INMARSAT M', category: 'yes_no' },
					{ key: 'inmarsat_mini', label: 'INMARSAT Mini-M', category: 'yes_no' },
					{ key: 'epirb_406_mhz', label: 'EPIRB 406 MHz', category: 'yes_no' },
					{ key: 'epirb_121_5_mhz', label: 'EPIRB 121.5 MHz', category: 'yes_no' },
					{ key: 'epirb_identification_code', label: 'EPIRB ID Code' },
					{ key: 'sart', label: 'SART', category: 'yes_no' },
					{ key: 'raft_count', label: 'Life Raft Count' },
					{ key: 'lifeboat_count', label: 'Lifeboat Count' },
					{ key: 'vessel_capacity', label: 'Vessel Capacity' }
				]
			},
			{
				source: 'exemption_request',
				title: 'Exemption Request',
				fields: [
					{ key: 'ship_type', label: 'Ship Type', category: 'ship_type' },
					{ key: 'port_registry', label: 'Port of Registry' },
					{ key: 'owner', label: 'Owner' },
					{ key: 'operater', label: 'Operator' },
					{ key: 'charter', label: 'Charterer' },
					{ key: 'agent', label: 'Agent' },
					{ key: 'foreign_port', label: 'Foreign Port' },
					{
						key: 'radiotelephone_exempt_req',
						label: 'Radiotelephone Exemption',
						category: 'yes_no'
					},
					{ key: 'gmdss_exemp_req', label: 'GMDSS Exemption', category: 'yes_no' },
					{ key: 'radio_dir_exempt_req', label: 'Radio Direction Exemption', category: 'yes_no' },
					{ key: 'vessel_size_exempt', label: 'Vessel Size Exemption', category: 'yes_no' },
					{ key: 'equipment_exempt', label: 'Equipment Exemption', category: 'yes_no' },
					{ key: 'ltd_routes_exempt', label: 'Limited Routes Exemption', category: 'yes_no' },
					{ key: 'cond_voyages_exempt', label: 'Conditional Voyages Exemption', category: 'yes_no' },
					{ key: 'other_exempt', label: 'Other Exemption', category: 'yes_no' },
					{ key: 'other_exempt_desc', label: 'Other Exemption Description' },
					{ key: 'prev_exempt_file_number', label: 'Previous Exemption File Number' },
					{ key: 'number_of_crew', label: 'Number of Crew' },
					{ key: 'number_passengers', label: 'Number of Passengers' },
					{ key: 'number_others', label: 'Number of Others' }
				]
			},
			{
				// FCC splits one long voyage description across several
				// sequence-numbered rows; the API returns them in order and
				// this renders them re-joined as the single sentence they
				// were always meant to be.
				source: 'voyages',
				title: 'Voyage Description',
				kind: 'joined_text',
				joinField: 'voyage_description'
			}
		]
	}
};

export const PERSONAL_SERVICE_LIST = Object.values(PERSONAL_SERVICES);

/** Route for a search result / related-identity row, by its `source`. */
export function serviceRoute(source, key) {
	if (source === 'amateur') return `/amateur/${key}`;
	if (source === 'tower') return `/towers/${key}`;
	if (PERSONAL_SERVICES[source]) return `${PERSONAL_SERVICES[source].route}/${key}`;
	return null;
}
